pkg/network/discovery.go

```diff
@@ -203,31 +216,53 @@ func (d *DefaultDiscovery) GoodPeers() []AddressWithCapabilities {
 	return addrs
 }
 
-// RegisterGoodAddr registers a known good connected address that has passed
+// RegisterGood registers a known good connected peer that has passed
 // handshake successfully.
-func (d *DefaultDiscovery) RegisterGoodAddr(s string, c capability.Capabilities) {
+func (d *DefaultDiscovery) RegisterGood(p AddressablePeer) {
+	var s = p.PeerAddr().String()
 	d.lock.Lock()
-	d.goodAddrs[s] = c
+	d.handshakedAddrs[s] = true
+	d.goodAddrs[s] = p.Version().Capabilities
 	delete(d.badAddrs, s)
```

Comment:
Don't we need to remove it from connected as far?

Reply:
That's intentional. We have this connection, so it's OK to keep it. Unregistration will kick it out.